src/IceRpc.Protobuf/MessageParserExtensions.cs

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) ZeroC, Inc.
+
+using Google.Protobuf;
+using System.Buffers;
+using System.Buffers.Binary;
+using System.Diagnostics;
+using System.IO.Pipelines;
+
+namespace IceRpc.Protobuf;
+
+/// <summary>Provides an extension method for <see cref="MessageParser{T}" />.</summary>
+public static class MessageParserExtensions
+{
+    /// <summary>Decodes a Protobuf length prefixed message from a <see cref="PipeReader" />.</summary>
+    /// <param name="parser">The <see cref="MessageParser{T}" /> used to parse the message data.</param>
+    /// <param name="reader">The <see cref="PipeReader" /> containing the Protobuf length prefixed message.</param>
+    /// <param name="maxMessageLength">The maximum allowed length.</param>
+    /// <param name="cancellationToken">A cancellation token that receives the cancellation requests.</param>
+    /// <returns>The decoded message object.</returns>
+    public static async ValueTask<T> DecodeFromLengthPrefixedMessageAsync<T>(
+        this MessageParser<T> parser,
+        PipeReader reader,
+        int maxMessageLength,
+        CancellationToken cancellationToken) where T : IMessage<T>
+    {
+        ReadResult readResult = await reader.ReadAtLeastAsync(5, cancellationToken).ConfigureAwait(false);
+        if (readResult.Buffer.Length < 5)
+        {
+            reader.AdvanceTo(readResult.Buffer.End);
+            throw new InvalidDataException(
+                $"The payload has {readResult.Buffer.Length} bytes, but 5 bytes were expected.");
+        }
+
+        if (readResult.Buffer.FirstSpan[0] == 1)
+        {
+            reader.AdvanceTo(readResult.Buffer.GetPosition(5));
+            throw new NotSupportedException("IceRPC does not support Protobuf compressed messages");
+        }
+        int messageLength = DecodeMessageLength(readResult.Buffer.Slice(1, 4));
+        reader.AdvanceTo(readResult.Buffer.GetPosition(5));
+        if (messageLength >= maxMessageLength)
+        {
+            throw new InvalidDataException("The message length exceeds the maximum value.");
+        }
+
+        readResult = await reader.ReadAtLeastAsync(messageLength, cancellationToken).ConfigureAwait(false);
+        if (readResult.Buffer.Length < messageLength)
+        {
+            reader.AdvanceTo(readResult.Buffer.End);
+            throw new InvalidDataException(
+                $"The payload has {readResult.Buffer.Length} bytes, but {messageLength} bytes were expected.");
+        }
+
+        T message = parser.ParseFrom(readResult.Buffer.Slice(messageLength));
```

Comment:
Does ParseFrom check it read all the bytes? 

Assume it does, please add a comment.

In the equivalent Slice code, we have 
https://github.com/icerpc/icerpc-csharp/blob/ecbe38c0fd584046384b0589b9c94b278fb5ca6f/src/IceRpc.Slice/Internal/IncomingFrameExtensions.cs#L53

Reply:
The parse ends with a call to `CheckReadEndOfStreamTag`

```
        /// <summary>
        /// Verifies that the last call to ReadTag() returned tag 0 - in other words,
        /// we've reached the end of the stream when we expected to.
        /// </summary>
        /// <exception cref="InvalidProtocolBufferException">The 
        /// tag read was not the one specified</exception>
        public static void CheckReadEndOfStreamTag(ref ParserInternalState state)
        {
            if (state.lastTag != 0)
            {
                throw InvalidProtocolBufferException.MoreDataAvailable();
            }
        }
```

This seems more like a tag end marked, than an end of buffer check.

Reply:
When using the ReadOnlySequence overload, we don't know how many bytes the ParseFrom method read.

Reply:
Maybe just add a TODO for now?